### PR TITLE
bump github.com/Azure/azure-service-operator/v2 from v2.3.0 to v2.4.0

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**: Bumps ASO to v2.4.0.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

~~ASO v2.4 is NOT published yet: https://github.com/Azure/azure-service-operator/releases/tag/v2.4.0~~
ASO v2.4 is live!
- https://github.com/Azure/azure-service-operator/releases/tag/v2.4.0

This PR is split into two commits: The first is the intended set of changes that should be merged. The second is some hacks to refer to custom built artifacts to get test signal before the official ASO manifests/images are published. The second commit MUST be dropped before this is merged. The only other modification to be made is to update go.mod to refer to the real v2.4.0 of ASO instead of a commit on its `main` branch: `go get github.com/Azure/azure-service-operator/v2@v2.4.0`.
- [x] done

/hold for the above changes

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
CAPZ now depends on ASO v2.4.0
```

### CODE REVIEW

File: main.go
Comment (nojnhuh) [inline main.go:566]: This hunk is all copy-pasta from CAPI 1.6 so we don't have to update our CAPI.

File: main.go
Comment (nojnhuh) [inline main.go:95]: These changes are mirrored from #4182 to account for breaking changes in controller-runtime 0.16, which ASO 2.4 has a hard dependency on.

File: main.go
Comment (nojnhuh) [inline main.go:255]: The linter was throwing an error since this pulls in a new version of `klog` which deprecates `klogr.New` so that caused a little bit of extra collateral damage.

**Comment by nojnhuh** [comment]
> Dependency review is failing because of this [GHSA-8pgv-569h-w5rw](https://github.com/advisories/GHSA-8pgv-569h-w5rw) which I think needs #4261 to merge first otherwise I see build failures trying to update only that module.
> 
> All the e2e-like tests seem like flakes.

Once #4261 merges, we should be able to rebase this branch and then run `go get go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@latest` to update fix the CVE.

**Comment by nojnhuh** [comment]
> Once #4261 merges, we should be able to rebase this branch and then run `go get go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@latest` to fix the CVE.

Done

File: go.mod
Comment (willie-yao) [inline go.mod:18]: Should this be changed to v2.4.0?

File: main.go
Comment (willie-yao) [inline main.go:274]: Can this if statement be combined with the one on line 257?

File: go.mod
Comment (jackfrancis) [reply go.mod:18]: indeed!

Comment (jackfrancis) [reply go.mod]: done, thx

**Review by CecileRobertMichon — COMMENTED** [review]
@jackfrancis can you please squash and remove the [WIP] title prefix when this is ready for review?

File: go.mod
Comment (CecileRobertMichon) [inline go.mod:58]: is it ok to make this change before bumping CAPI to 1.6?

Comment (jackfrancis) [reply go.mod:58]: I believe this is a ASO v2.4.0 requirement.

I don't think there would be a strict requirement of staying in sync w/ CAPI (though certainly a best practice). If there were k8s breaking API ramifications our E2E tests would tell us that. Do we have a documented contract to stick to the same version as CAPI's latest version in CAPZ (and other providers)?

Comment (CecileRobertMichon) [reply go.mod:58]: I think you're right

let's ask in the CAPI office hours tomorrow am to get confirmation?

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve
/hold for confimation from CAPI that it's ok to bump controller runtime

File: go.mod
Comment (jackfrancis) [reply go.mod:58]: Confirmed in CAPI this is not a strict requirement

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/hold cancel

**Review by jackfrancis — APPROVED** [review]
/lgtm

**Pull request merged**
